discourse: Add package and NixOS module

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Package [Discourse](https://www.discourse.org/) and add a NixOS module.

Also, update `v8` to version `8.4.255`, which was needed by the `libv8` and `mini_racer` gems.

###### Todo

- [x] Write manual entry
- [x] Write NixOS test

###### Nice to have

- [x] Handling of incoming email
- [x] Updater script

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

talyz committed:
v8: 7.4.255 -> 8.4.255

talyz committed:
discourse: init at 2.6.3

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/default.nix:
```diff
@@ -0,0 +1,167 @@
+      export RAILS_ENV=production
+
+      bundle exec rake db:migrate >/dev/null
+      rm -rf tmp/*
```
```suggestion
      rm -r tmp/*
```

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/default.nix:
```diff
@@ -0,0 +1,167 @@
+
+    mkdir -p $out/share
+    cp -r . $out/share/discourse
+    rm -rf $out/share/discourse/log
```
```suggestion
    rm -r $out/share/discourse/log
```

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/default.nix:
```diff
@@ -0,0 +1,167 @@
+    homepage = "https://www.discourse.org/";
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ talyz ];
+    license = licenses.gpl2;
```
gpl2Only or gpl2Plus?

talyz replied:
Ah, true. It's `or later`.

SuperSandro2000 reviewed (commented):
pkgs/servers/web-apps/discourse/disable_jhead.patch is an empty file.

talyz commented:
@SuperSandro2000 Good catch! Strange, though. Oh well, it's fixed now, anyhow.

talyz committed:
nixos/discourse: Add test

talyz committed:
discourse: Add updater script

talyz committed:
nixos/discourse: Add NixOS manual entry

talyz commented:
I think this is ready now (except for squashing all the fixups). A summary of changes made since opening the PR:

- Added a basic NixOS test which makes sure logging in as the admin user works
- Added an update script which also prints a diff of files used for the module
- Wrote a NixOS manual entry
- Switched off `DynamicUser`, which was necessary to let nginx access persistent state
- Support ACME
- Optimize the nginx config

..and other small fixes.

aanderse reviewed (approved):
Awesome work @talyz... as always!

A few points I'll bring up, take them or leave them:
- Personally I prefer to use the [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) suggested route of `settings` over `extraConfig`, even the way you have implemented it. Suggested change [here](https://github.com/NixOS/nixpkgs/commit/6fbc6aba3891b76ae750537097545ff35d068aea).
- Inbound e-mail should be **super** simple to setup with `postfix` - no mucking around with POP because it looks like `discourse` exposes an api endpoint you could just throw inbound e-mail at via curl. Might be worth a mention in the docs, or a future PR :man_shrugging: I only mention because e-mail seems popular on our NixOS discourse.

Again, awesome work, good to go in my opinion. Any additional changes are gravy on top :+1:

talyz commented:
@aanderse Thanks! Good points :+1: 

- I've renamed `extraConfig` to `settings` as suggested. I couldn't use your suggested change because it always prioritizes the settings in the `config` section for the merge, so I went with a simple name change instead.
- Yeah, it might be simple - I didn't look too deep into it. It just looked trickier than POP3 because they have a [separate service](https://github.com/discourse/mail-receiver) handling it and don't implement it in the docker container since ["it is very tricky to set up correctly"](https://github.com/discourse/discourse_docker#email) :sweat_smile: I can have a look at it, though - it would definitely be a nice addition :slightly_smiling_face:

aanderse commented:
>     * I couldn't use your suggested change because it always prioritizes the settings in the `config` section for the merge

I believe the intention of RFC42 is to use `mkDefault` and `mkForce` to handle this. If you want a setting to be easily overridden by the sysadmin without second consideration then use `mkDefault`. If you want to imply a value shouldn't be changed unless the sysadmin knows what they are doing then they can use `mkForce`.

```
services.discourse.settings = mapAttrs (name: mkDefault) {
# all your settings here
}:
```

This would be the equivalent to what you have, I believe. Either way is completely fine 👍 Thanks @talyz!

I'll send you some config for `postfix` shortly that will probably work.

talyz committed:
discourse-mail-receiver: Init at 4.0.7

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/mail_receiver/default.nix:
```diff
@@ -0,0 +1,30 @@
+    wrapProgram $out/bin/receive-mail --set RUBYLIB $out/lib
+    wrapProgram $out/bin/discourse-smtp-fast-rejection --set RUBYLIB $out/lib
+  '';
+}
```
Missing meta.

talyz replied:
Oops! I've added it, but upstream doesn't list a license. I've created https://github.com/discourse/mail-receiver/pull/8 to remedy this, but for now I've left out the license attribute.

talyz committed:
rl-2105: Add Discourse module note

talyz commented:
Work done since the last update:
- Implemented support for incoming email, using Discourse's [mail-receiver](https://github.com/discourse/mail-receiver)
- Implemented support for declarative site setting defaults, i.e. defining the default values for the settings you can change through the admin UI
- Support arbitrary secrets for site settings
- Change the `settings`-options to be more in line with [RFC 0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)
- Test outgoing and incoming email in the test
- Update the manual entry
- Added to release notes

This is pretty much all I can think of to implement, so I think this is done and ready to merge now.

talyz commented:
Running the test on slow machines, one of the requests timed out. I've now made Unicorn's request timeout configurable and set it to `900` in the test.

talyz commented:
I plan to merge this on Friday unless someone objects.

NinjaTrappeur commented on pkgs/development/libraries/v8/default.nix:
```diff
@@ -39,29 +39,31 @@ let
-      sha256 = "02hczcg43m36q8j1kv5j3hq9czj9niiil9w13w22vzv2f3c67dvn";
-    })
     ./darwin.patch
+    ./gcc_arm.patch  # Fix building zlib with gcc on aarch64, from https://gist.github.com/Adenilson/d973b6fd96c7709d33ddf08cf1dcb149
```
Is this tracked upstream?

talyz replied:
Well, this applies specifically to Google's zlib fork. It's fixed in a newer version of it ([here](https://chromium.googlesource.com/chromium/src/third_party/zlib/+/f87c8e28893b1eb7cacf8043200f8daaef21a680%5E%21/#F0)), so going to an even newer version of `v8` would likely deprecate this patch.

I tried going to the latest version of `v8` available, but ran into other build errors, so since upgrading `v8` wasn't really the main intention of the PR, I settled for this one to save time.

flokli commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+{
+  options = {
+    services.discourse = {
+      enable = lib.mkOption {
```
mkEnableOption

flokli commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+        '';
+      };
+
+      user = lib.mkOption {
+        type = lib.types.str;
+        default = "discourse";
+        description = ''
+          User account under which Discourse runs.
+        '';
+      };
+
+      group = lib.mkOption {
+        type = lib.types.str;
+        default = "discourse";
+        description = ''
+          Group account under which Discourse runs.
+        '';
+      };
```
Do we /need/ this to be configurable? I assume we can only set up one instance of this anyways…

talyz replied:
No, not really. I've removed it now, which simplifies things a bit :)

flokli commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+
+      hostname = lib.mkOption {
+        type = lib.types.str;
+        default = config.networking.hostName;
```
```suggestion
        default = config.networking.fqdn;
        defaultText = "config.networking.fqdn";
```

talyz replied:
Yep, that's more reasonable. It required some working around `config.networking.domain` being `null`, but is definitely more what we want.

flokli commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+        '';
+      };
+
+      sslCertificate = lib.mkOption {
+        type = with lib.types; nullOr path;
+        default = null;
+        example = "/run/keys/ssl.cert";
+        description = ''
+          The path to the server SSL certificate. Set this to enable
+          SSL.
+        '';
+      };
+
+      sslCertificateKey = lib.mkOption {
+        type = with lib.types; nullOr path;
+        default = null;
+        example = "/run/keys/ssl.key";
+        description = ''
+          The path to the server SSL certificate key. Set this to
+          enable SSL.
+        '';
+      };
+
+      enableACME = lib.mkOption {
+        type = lib.types.bool;
+        default = cfg.sslCertificate == null && cfg.sslCertificateKey == null;
+        description = ''
+          Whether an ACME certificate should be used to secure
+          connections to the server.
+        '';
+      };
```
While it's in the defaults, ACME/non-ACME-behaviour should be visible just from exploring the NixOS options.

talyz replied:
Added a `defaultText` to clarify.

flokli commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+
+    services.nginx = lib.mkIf cfg.nginx.enable {
+      enable = true;
+      package = cfg.nginx.package.override {
+        modules = cfg.nginx.package.modules ++ [ pkgs.nginxModules.brotli ];
+      };
```
This globally sets `services.nginx.package` (and exposes an option for it too). Is there a way we can expose which nginx modules are enabled in the nginx module, and just have this module configuring that?

talyz replied:
Indeed, that's what I wanted to begin with, but at the time it felt a bit out of scope for this. I can look into it now, though - it would be a much nicer solution.

ryantm replied:
I think you're right that fixing this is scope creep. It is behind the Discourse enable flag, so it should be safe to leave it as is and improve it later.

talyz replied:
Well, I've added a pretty simple fix for this now, so please take a look.

ryantm replied:
Looks good to me

NinjaTrappeur commented:
I'm interested by this module as well. I'm looking forward getting this to NixPkgs.

----------------------------------

Apart from Flokli's comments, I was also a bit concerned about the v8 major version bump. I made a quick search in NixPkgs to see what's depending on that:

```
- proglodyte-wasm => seems abandonned
- rq => active-ish, last updated in february 2020
- There's a v8 R module
- There's a libv8 ruby gem
- plv8 => active-ish, last updated in July 2020
```

I don't see any widely used dependency here. Worst case scenario we'll break a leaf.

NinjaTrappeur commented on nixos/modules/services/web-apps/discourse.nix:
```diff
@@ -0,0 +1,1079 @@
+    mkdir -p $out/bin
+    makeWrapper ${cfg.package.rubyEnv}/bin/rake $out/bin/discourse-rake \
+        ${lib.concatStrings (lib.mapAttrsToList (name: value: "--set ${name} '${value}' ") discourseEnv)} \
+        --prefix PATH : ${lib.makeBinPath discourseDeps} \
```
Is there a particular thing preventing us from wrapping the binary with these dependencies directly in the discourse package fixupPhase?

talyz replied:
It's not really dependent on any runtime configuration, not anymore (and rake shouldn't need access to the `UNICORN_`-variables), so I moved `discourse-rake` into the package instead. Note that it uses `rake` from the `rubyEnv`, though, so it's not possible to do in the `fixupPhase`.

ryantm commented:
Thank you, this looks like it was a lot of work to set up! I strongly urge you to reconsider this packaging method. I have used Discourse for many years, and I see they are often changing how their system works quite dramatically, so I expect this packaging method to be a lot of work to maintain.

I think a simpler and more maintainable route is to use the NixOS module system to set up the Docker container provided by upstream.

That said, please don't view this is as me trying to veto inclusion of this particular packaging method, just a warning that you might be getting yourself into a lot of work maintaining it!

ryantm commented:
I read through the code and what you are doing here is not as bad as I expected! I didn't see any mentions of Discourse plugins. Is there some mechanism for adding them that I missed?

talyz commented:
@NinjaTrappeur Yeah, it seems to have surprisingly few dependents, many of which are already broken or abandoned. I did try building all of them and the build status was unchanged for all packages. `plv8` even has its own version of `v8` already, so it shouldn't be affected.

talyz commented:
> Thank you, this looks like it was a lot of work to set up! I strongly urge you to reconsider this packaging method. I have used Discourse for many years, and I see they are often changing how their system works quite dramatically, so I expect this packaging method to be a lot of work to maintain.

Yes, it was a lot of work to get everything in place, but I think the most difficult parts are unlikely to change drastically. For the things that need regular upkeep, I implemented some helpers in the update script. The test should also, hopefully, be extensive enough to provide assurance when updating, at least for basic functionality and email.

> I think a simpler and more maintainable route is to use the NixOS module system to set up the Docker container provided by upstream.

While using the docker container is certainly an option, I think packaging it natively provides much more flexibility and customizability. Also, while probably a minor thing, it also runs on `aarch64`. If it turns out this is completely unmaintainable, we could switch to the container, but it would likely not be able to offer much of this module's functionality.

> I read through the code and what you are doing here is not as bad as I expected! I didn't see any mentions of Discourse plugins. Is there some mechanism for adding them that I missed?

That's great! I must admit that I completely forgot about plugins. I've added support now, via the `services.discourse.plugins` option. Thanks for the reminder!

talyz committed:
nixos/nginx: Add an option to specify additional third-party modules

talyz committed:
nixos/discourse: Init

talyz commented:
My new plan is to merge this on Thursday unless there are objections.

picnoir reviewed (approved):
No objections here.

I'm not 100% sure hydra will enjoy this pretty crazy (but apparently necessary) buildPhase. If it doesn't, there's no dependency on this package elsewhere, it won't be a big deal. We'll move on from there.

ajs124 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -397,13 +397,27 @@ in
         defaultText = "pkgs.nginxStable";
         type = types.package;
+        apply = p: p.override {
+          modules = p.modules ++ cfg.additionalModules;
```
I'd really appreciate if people ran all `nginx*` nixos tests when changing the module.
This broke the `nginx-variants` test again, because `tengine` doesn't have `.modules`. I don't use `tengine` and maybe nobody does, but the test still seems useful, because it tests some other nginx variants.

talyz replied:
Fixed in #119427.

dpausp commented:
Thanks a lot. I've been running some Discourse instances but I never liked their special docker-based approach (ok, they started very early). I'll try out the module. Maybe we can host the NixOS Discourse on NixOS in the future? :)

talyz commented:
@dpausp Thanks for trying it out! Yeah, maybe! @zimbatm probably knows better whether that's feasible.

zimbatm commented:
Yes, technically it's possible. It would add more work to the infra team (Eelco, Graham and a tiny bit of me) so I would prefer to avoid it. We did have to upgrade from the free hosted plan so it's starting to become more attractive though.

The best thing you or somebody else could do is to run an instance of your own and gather notes on all the administrative knowledge that is needed to keep it up and running. How often are there upgrades, how hard is it to keep up to date, how do you handle backup and restore, ...

talyz commented:
@zimbatm Sure! If everything goes to plan, I'll be hosting an instance for my company using this module in the near future. So, if there's interest in switching to using it, I'll hopefully be able to give pointers / help out.

zimbatm commented:
Sounds great! It looks like we might need that sooner than I hoped: https://discourse.nixos.org/t/discourse-not-sending-some-emails/12125/26?u=zimbatm

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/discourse-not-sending-some-emails/12125/28
